Chart: Add `controller.progressDeadlineSeconds`.

TrondT opened the pull request:
Add option to set progressDeadlineSeconds on Deployment: [progressDeadlineSeconds](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#failed-deployment)

## What this PR does / why we need it:
Some users would like to detect a failing deployment quicker than the default 600 seconds.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
Tested with helm template with and without progressDeadlineSeconds set

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.

Gacko commented:
/triage accepted
/kind feature
/priority backlog
/hold

Hello and thank you for your contribution! Some notes: Please...

- ... also implement this for the `DaemonSet`.
- ... implement Helm Unit Tests for the both the `Deployment` and the `DaemonSet`.
- Do not change the `go.sum`.

Thank you in advance!

TrondT commented:
Thanks for your quick response.
Will update accordingly.

TrondT commented:
Regarding your comments, I have some thoughts to consider:
... set a non-nil default value.
By using 'nil' we will assume the kubernets default, as described here: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#deployment-v1-apps
I believe a good practice is to not redefine if there are no need for such.
... also implement this for the DaemonSet.
DaemonSet spec does not include `progressDeadlineSeconds`, ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#daemonset-v1-apps
... implement Helm Unit Tests for the both the Deployment and the DaemonSet.
Implemented test for Deployment.
... use helm-docs for generating the README.md.
Done.
... do not change the go.sum.
Reverted.

TrondT commented:
I'll be happy to set default value for progressDeadlineSeconds to 600 if you thing that is a better solution... 
Just let me know.

Gacko commented:
Can you try `0` as default?

Gacko commented:
I think we had a misunderstanding here: Using `0` in conjunction with an `if` would only render it if the value is not `0`. So if someone does not change the default of `0`, this field would never get rendered. But if you change it to, let's say, 30, then it gets rendered as expected.

TrondT commented:
Now used 0 as default, so with this input:
```
spec:
  progressDeadlineSeconds: 0
```
the Deployment is rendered as follows:
```
spec:
  progressDeadlineSeconds: 600
```
  Also tested in our cluster.

Gacko commented on charts/ingress-nginx/templates/controller-deployment.yaml:
```diff
@@ -22,6 +22,7 @@ spec:
   replicas: {{ .Values.controller.replicaCount }}
   {{- end }}
   revisionHistoryLimit: {{ .Values.revisionHistoryLimit }}
+  progressDeadlineSeconds: {{ .Values.controller.progressDeadlineSeconds }}
```
```suggestion
  {{- if .Values.controller.progressDeadlineSeconds }}
  progressDeadlineSeconds: {{ .Values.controller.progressDeadlineSeconds }}
  {{- end }}
```

Gacko commented on charts/ingress-nginx/tests/controller-deployment_test.yaml:
```diff
@@ -177,3 +177,17 @@ tests:
           path: spec.template.spec.containers[0].image
           value: registry.k8s.io/ingress-nginx/controller:my-little-custom-tag@sha256:faa2d18687f734994b6bd9e309e7a73852a81c30e1b8f63165fcd4f0a087e3cd
+
+  - it: should create a Deployment with default `spec.progressDeadlineSeconds`
+    asserts:
+      - equal:
+          path: spec.progressDeadlineSeconds
+          value: 600
```
You can omit this.

Gacko commented on charts/ingress-nginx/tests/controller-deployment_test.yaml:
```diff
@@ -177,3 +177,17 @@ tests:
+          path: spec.progressDeadlineSeconds
+          value: 600
+
+  - it: should create a Deployment with `spec.progressDeadlineSeconds` set to custom value
```
```suggestion
  - it: should create a Deployment with `progressDeadlineSeconds` if `controller.progressDeadlineSeconds` is set
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -56,6 +56,9 @@ controller:
   proxySetHeaders: {}
   # -- Will add custom headers before sending response traffic to the client according to: https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#add-headers
   addHeaders: {}
+  # -- (int) Optionally set the deployment progressDeadlineSeconds. Note that the default 0 value will use the kubernetes default.
+  # Ref.: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#deployment-v1-apps
```
```suggestion
  # -- Specifies the number of seconds you want to wait for the controller to progress before the system reports back that the deployment has failed progressing.
  # Ref.: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -56,6 +56,9 @@ controller:
   addHeaders: {}
+  # -- (int) Optionally set the deployment progressDeadlineSeconds. Note that the default 0 value will use the kubernetes default.
+  # Ref.: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#deployment-v1-apps
+  progressDeadlineSeconds: 600
```
```suggestion
  progressDeadlineSeconds: 0
```

Gacko reviewed (approved):
/unhold
/lgtm
